extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java

```diff
@@ -165,26 +152,56 @@ public void logCleanup(BuildProducer<LogCleanupFilterBuildItem> logCleanupFilter
+        JandexBeanInfoAdapter beanInfoAdapter = new JandexBeanInfoAdapter(index);
+        JandexMemberInfoAdapter memberInfoAdapter = new JandexMemberInfoAdapter(index);
+
+        List<MethodInfo> collectedMetricsMethods = new ArrayList<>();
+        Map<DotName, ClassInfo> collectedMetricsClasses = new HashMap<>();
+
+        for (DotName metricAnnotation : METRICS_ANNOTATIONS) {
+            Collection<AnnotationInstance> metricAnnotationInstances = index.getAnnotations(metricAnnotation);
+            for (AnnotationInstance metricAnnotationInstance : metricAnnotationInstances) {
+                AnnotationTarget metricAnnotationTarget = metricAnnotationInstance.target();
+                switch (metricAnnotationTarget.kind()) {
+                    case METHOD: {
+                        MethodInfo method = metricAnnotationTarget.asMethod();
+                        if (!collectedMetricsMethods.contains(method)) {
+                            collectedMetricsMethods.add(method);
+                        }
+                        break;
+                    }
+                    case CLASS: {
+                        ClassInfo clazz = metricAnnotationTarget.asClass();
+                        collectMetricsClassAndSubClasses(index, collectedMetricsClasses, clazz);
+                        break;
+                    }
+                }
             }
+        }
+
+        for (ClassInfo clazz : collectedMetricsClasses.values()) {
             BeanInfo beanInfo = beanInfoAdapter.convert(clazz);
             for (MethodInfo method : clazz.methods()) {
-                if (containsMetricsAnnotations(method.annotations()) ||
-                        containsMetricsAnnotations(clazz.classAnnotations())) {
-                    metrics.registerMetrics(beanInfo, memberInfoAdapter.convert(method));
-                }
+                metrics.registerMetrics(beanInfo, memberInfoAdapter.convert(method));
+            }
+        }
+
+        for (MethodInfo method : collectedMetricsMethods) {
+            ClassInfo declaringClazz = method.declaringClass();
+            if (!collectedMetricsClasses.containsKey(declaringClazz.name())) {
+                BeanInfo beanInfo = beanInfoAdapter.convert(declaringClazz);
+                metrics.registerMetrics(beanInfo, memberInfoAdapter.convert(method));
             }
         }
     }
 
-    private boolean containsMetricsAnnotations(Collection<AnnotationInstance> annotations) {
-        return annotations.stream().anyMatch(annotation -> metricsAnnotations.contains(annotation.name()));
+    private void collectMetricsClassAndSubClasses(IndexView index, Map<DotName, ClassInfo> collectedMetricsClasses,
+            ClassInfo clazz) {
+        collectedMetricsClasses.put(clazz.name(), clazz);
+        for (ClassInfo subClass : index.getAllKnownSubclasses(clazz.name())) {
+            collectMetricsClassAndSubClasses(index, collectedMetricsClasses, subClass);
```

Comment:
I think you should just do:
```
collectedMetricsClasses.put(subClass.name(), subClass);
```
You don't need to go recursive as `index.getAllKnownSubclasses(clazz.name())` will return all the subclasses, not just the direct ones.

Reply:
Good to know. Thanks.